src/VisualStudio/Core/Impl/SolutionExplorer/AnalyzerItem/AnalyzerItemSource.cs

```diff
@@ -2,237 +2,171 @@
-                return;
-            }
+    private readonly BulkObservableCollection<AnalyzerItem> _items = [];
 
-            var project = _analyzersFolder.Workspace
-                            .CurrentSolution
-                            .GetProject(_analyzersFolder.ProjectId);
+    private readonly CancellationTokenSource _cancellationTokenSource = new();
+    private readonly AsyncBatchingWorkQueue _workQueue;
 
-            if (project != null &&
-                project.AnalyzerReferences != _analyzerReferences)
-            {
-                _analyzerReferences = project.AnalyzerReferences;
+    private IReadOnlyCollection<AnalyzerReference>? _analyzerReferences;
 
-                _analyzerItems.BeginBulkOperation();
+    private Workspace Workspace => _analyzersFolder.Workspace;
+    private ProjectId ProjectId => _analyzersFolder.ProjectId;
 
-                var itemsToRemove = _analyzerItems
-                                        .Where(item => !_analyzerReferences.Contains(item.AnalyzerReference))
-                                        .ToArray();
+    public AnalyzerItemSource(
+        AnalyzersFolderItem analyzersFolder,
+        IAnalyzersCommandHandler commandHandler,
+        IAsynchronousOperationListenerProvider listenerProvider)
+    {
+        _analyzersFolder = analyzersFolder;
+        _commandHandler = commandHandler;
 
-                var referencesToAdd = GetFilteredAnalyzers(_analyzerReferences, project)
-                                        .Where(r => !_analyzerItems.Any(item => item.AnalyzerReference == r))
-                                        .ToArray();
+        _workQueue = new AsyncBatchingWorkQueue(
+            DelayTimeSpan.Idle,
+            ProcessQueueAsync,
+            listenerProvider.GetListener(FeatureAttribute.SourceGenerators),
+            _cancellationTokenSource.Token);
 
-                foreach (var item in itemsToRemove)
-                {
-                    _analyzerItems.Remove(item);
-                }
+        this.Workspace.WorkspaceChanged += OnWorkspaceChanged;
 
-                foreach (var reference in referencesToAdd)
-                {
-                    _analyzerItems.Add(new AnalyzerItem(_analyzersFolder, reference, _commandHandler.AnalyzerContextMenuController));
-                }
+        // Kick off the initial work to determine the starting set of items.
+        _workQueue.AddWork();
+    }
 
-                var sorted = _analyzerItems.OrderBy(item => item.AnalyzerReference.Display).ToArray();
-                for (var i = 0; i < sorted.Length; i++)
-                {
-                    _analyzerItems.Move(_analyzerItems.IndexOf(sorted[i]), i);
-                }
+    public object SourceItem => _analyzersFolder;
 
-                _analyzerItems.EndBulkOperation();
+    // Defer actual determination and computation of the items until later.
+    public bool HasItems => !_cancellationTokenSource.IsCancellationRequested;
```

Comment:
>HasItems [](http://example.com/codeflow?start=16&length=8)

Is this used in solution explorer to determine if it is expandable?

Reply:
correct.  This matches what many of our SE nodes do though, which is to guess that they have children while computing asynchronously.